src/main/java/fi/metatavu/metamind/server/rest/StoriesApiImpl.java

```diff
@@ -266,45 +322,78 @@ public Response createSession(fi.metatavu.metamind.rest.model.Session body, UUID
+      Set<UUID> permittedUsersId = authenticationController.getResourcePermittedUsers(REALM_NAME, createdResourceId, String.format("Session-%s", session.getId()), SCOPES);
+      if (!permittedUsersId.contains(loggedUserId) || loggedUserId == null) {
+        return createBadRequest(String.format(NO_USER_FOUND_MESSAGE));
+      }
+    } catch (Exception e) {
+      logger.warn(String.format(EXCEPTION_CAUGHT, e.getMessage()));
+    }
+    
     return createOk(sessionTranslator.translateSession(session));
   }
 
   @Override
   public Response createStory(Story body) {
-    UUID loggedUserId = getLoggerUserId();
+
     Locale locale = LocaleUtils.toLocale(body.getLocale());
     
-    // TODO: Permission check
-
-    return createOk(storyTranslator.translateStory(storyController.createStory(locale, body.getName(), body.getDafaultHint(), loggedUserId)));
+    fi.metatavu.metamind.persistence.models.Story story = storyController.createStory(locale, body.getName(), body.getDafaultHint(), loggedUserId);
+    try {
+      UUID createdResourceId = authenticationController.createProtectedResource(loggedUserId, String.format(STORY_NAME_BEGINNING, story.getId()), String.format("/v2/stories/%s", story.getId()), "story", SCOPES);
+      List<UUID> policyIds = authenticationController.updatePermissionUsers(REALM_NAME);
+      authenticationController.upsertScopePermission(REALM_NAME, createdResourceId, SCOPES, String.format("Permission for story-%s", story.getId()), DecisionStrategy.AFFIRMATIVE, policyIds);
+      Set<UUID> permittedUsersId = authenticationController.getResourcePermittedUsers(REALM_NAME, createdResourceId, String.format(STORY_NAME_BEGINNING, story.getId()), SCOPES);
+      if (!permittedUsersId.contains(loggedUserId) || loggedUserId == null) {
+        return createBadRequest(String.format(NO_USER_FOUND_MESSAGE));
+      }
+    } catch (Exception e) {
+      logger.warn(String.format(EXCEPTION_CAUGHT, e.getMessage()));
+    }
+    
+    return createOk(storyTranslator.translateStory(story));
   }
 
   @Override
   public Response createVariable(Variable body, UUID storyId) {
-    UUID loggedUserId = getLoggerUserId();
-    
-    // TODO: Permission check
+
+    VariableType type = body.getType();
+    String validationScript = body.getValidationScript();
     
     fi.metatavu.metamind.persistence.models.Story story = storyController.findStoryById(storyId);
     if (story == null) {
       return createBadRequest(String.format("Story %s not found", storyId)); 
     }
     
+    fi.metatavu.metamind.persistence.models.Variable variable = storyController.createVariable(type, story, body.getName(), validationScript, loggedUserId);
+    
+    try {
+      UUID createdResourceId = authenticationController.createProtectedResource(loggedUserId, String.format(VARIABLE_NAME_BEGINNING, variable.getId()), String.format("/v2/stories/variables/%s", variable.getId()), "variable", SCOPES);
```

Comment:
Permissions should be defined only on story level, variable permissions are way too fine grained